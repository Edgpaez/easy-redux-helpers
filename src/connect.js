import { connect } from "react-redux";

const getWithDotNotation = (object, key) => {
  key = key.replace("state.", "");

  if (key.includes(".")) {
    let value = object;
    key.split(".").forEach(subkey => {
      value = value[subkey];
    });

    return value;
  }

  return object[key];
};

const buildMapStateToProps = propMap => (state, componentProps) => {
  const connectedProps = {};
  if (typeof propMap === "function") {
    return propMap(state, componentProps);
  }
  Object.keys(propMap).forEach(propName => {
    if (typeof propMap[propName] === "string") {
      // dot notation: state.potato.color
      try {
        connectedProps[propName] = getWithDotNotation(state, propMap[propName]);
      } catch (e) {
        throw new Error(
          `Error trying to map "${
            propMap[propName]
          }" to prop "${propName}", while connecting component to redux. ${
            propMap[propName]
          } is not defined`
        );
      }
    } else if (typeof propMap[propName] === "function") {
      // selectors
      connectedProps[propName] = propMap[propName].call(
        null,
        state,
        componentProps
      );
    }
  });
  return connectedProps;
};
const buildDispatchToProps = actions => dispatch => {
  const connectedActions = {};
  Object.keys(actions).forEach(key => {
    if (typeof actions[key] !== "function") {
      throw new Error(
        `withActions expects action creators. Received ${typeof actions[
          key
        ]} for ${key} action`
      );
    }

    connectedActions[key] = (...args) =>
      dispatch(actions[key].apply(null, args));
  });
  return { actions: connectedActions };
};

export default Component => {
  let stateToProps, dispatchToProps;
  return {
    to: config => {
      stateToProps = config["props"]
        ? buildMapStateToProps(config["props"])
        : undefined;
      dispatchToProps = config["actions"]
        ? buildDispatchToProps(config["actions"])
        : undefined;

      return connect(
        stateToProps,
        dispatchToProps
      )(Component);
    }
  };
};
