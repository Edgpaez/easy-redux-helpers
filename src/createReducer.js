import produce from "immer";

export default ({ initialState, ...reducers }) => {
  return (state = initialState, action) => {
    const srcReducer = reducers[action.type];

    return produce(state, draft => {
      srcReducer && srcReducer(draft, action);
    });
  };
};
