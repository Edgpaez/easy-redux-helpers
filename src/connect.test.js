import createReducer from "./createReducer";
import React from "react";
import { createStore } from "redux";
import Renderer from "react-test-renderer";
import { Provider } from "react-redux";
import connect from "./connect";
import { createSelector } from "reselect";

const config = {
  initialState: {
    todos: ["first"]
  },

  ADD: (state, action) => {
    state.todos.push(action.payload.text);
  }
};

const reducer = createReducer(config);
const store = createStore(reducer);
const getTotal = createSelector(
  state => state.todos,
  todos => todos.length
);

const BaseComponent = props => (
  <div>
    {props.todos.map((todo, index) => (
      <span key={index}>{todo}</span>
    ))}
    <span>Total: {props.total}</span>
    <button onClick={props.onAdd}>Add</button>
  </div>
);

const add_todo = text => ({ type: "ADD", payload: { text } });

it("connects state and selectors", () => {
  const ConnectedComponent = connect(BaseComponent).to({
    props: {
      todos: "state.todos",
      total: getTotal
    }
  });

  const tree = Renderer.create(
    <Provider store={store}>
      <ConnectedComponent />
    </Provider>
  );
  expect(tree.toJSON()).toMatchSnapshot();
});

it("connects actions", () => {
  const ConnectedComponent = connect(BaseComponent).to({
    props: {
      todos: "state.todos",
      total: getTotal
    },
    actions: {
      onAdd: add_todo
    }
  });

  const tree = Renderer.create(
    <Provider store={store}>
      <ConnectedComponent />
    </Provider>
  );

  const button = tree.root.findByType("button");
  button.props.onClick("second");
  button.props.onClick("third");
  expect(tree.toJSON()).toMatchSnapshot();
});

it("throws if the state you try to connect is is not defined", () => {
  const ConnectedComponent = connect(BaseComponent).to({
    props: {
      todos: "state.potato.potato"
    }
  });

  expect(() =>
    Renderer.create(
      <Provider store={store}>
        <ConnectedComponent />
      </Provider>
    )
  ).toThrowError(
    `Error trying to map "state.potato.potato" to prop "todos", while connecting component to redux. state.potato.potato is not defined`
  );
});

it("throws if you don't pass an action creator", () => {
  const ConnectedComponent = connect(BaseComponent).to({
    props: {
      todos: "state.todos"
    },
    actions: {
      onAdd: { type: "ADD" }
    }
  });

  expect(() =>
    Renderer.create(
      <Provider store={store}>
        <ConnectedComponent />
      </Provider>
    )
  ).toThrowError(
    `withActions expects action creators. Received object for onAdd action`
  );
});
