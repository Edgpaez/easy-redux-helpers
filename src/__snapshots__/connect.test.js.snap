// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can receive a callback to map state to props 1`] = `
<div>
  <span>
    first
  </span>
  <span>
    Total: 
    1
  </span>
  <button>
    Add
  </button>
</div>
`;

exports[`connects actions 1`] = `
<div>
  <span>
    first
  </span>
  <span>
    second
  </span>
  <span>
    third
  </span>
  <span>
    Total: 
    3
  </span>
  <button
    onClick={[Function]}
  >
    Add
  </button>
</div>
`;

exports[`connects state and selectors 1`] = `
<div>
  <span>
    first
  </span>
  <span>
    Total: 
    1
  </span>
  <button>
    Add
  </button>
</div>
`;
